File: day3/index1.js
const upSquare = n => Math.ceil(Math.sqrt(n));

const getLayer = n => {
  const m = upSquare(n);
  return Math.ceil((m - 1) / 2);
};

const N = 277678;

const dist = n => {
  if (n === 1) return 0;
  const l = getLayer(n);
  n -= (2 * l - 1) ** 2 + 1;
  const layerLen = 2 * l;
  const lower = layerLen * Math.floor(n / layerLen);
  const upper = lower + layerLen - 1;
  const mid = (lower + upper - 1) / 2;
  return l + Math.abs(n - mid);
};

console.log(dist(N));

// 1 8 16 24

File: clojure/day03/index1.clj
(defn ceil [n] (int (Math/ceil n)))
(defn floor [n] (int (Math/floor n)))

(defn up-square [n] (ceil (Math/sqrt n)))

(defn get-layer [n] 
  (let [m (up-square n)]
  (ceil (/ (dec m) 2))))

(def N 277678)

(defn sq [x] (* x x))

(defn dist [n] 
  (if (> n 1) (
    let [
      l (get-layer n)
      layer-len (* 2 l)
      n' (- n (inc (sq (dec layer-len))))
      lower (* layer-len (floor (/ n' layer-len)))
      upper (dec (+ lower layer-len))
      mid (/ (dec (+ lower upper)) 2)
    ]
      (+ l (Math/abs (- n' mid)))
    
  ) 0))

(println (dist N))
